java-frontend/src/main/java/org/sonar/java/ast/visitors/ExecutableLinesVisitor.java

```diff
@@ -0,0 +1,269 @@
+          addLine(closeBraceToken);
+        }
+      }
+    }
+
+    @Override
+    public void visitSynchronizedStatement(SynchronizedStatementTree tree) {
+      addLine(tree.synchronizedKeyword());
+      addLineOfCloseBrace(tree.synchronizedKeyword(), tree.block());
+      scan(tree.block());
+    }
+
+    @Override
+    public void visitSwitchStatement(SwitchStatementTree tree) {
+      addLines(tree.switchKeyword(), tree.openBraceToken());
+      scan(tree.cases());
+      addLine(tree.closeBraceToken());
+    }
+
+    @Override
+    public void visitVariable(VariableTree tree) {
+      SyntaxToken endToken = tree.endToken();
+      if (endToken != null && ";".equals(endToken.text())) {
+        addLine(endToken);
+      }
+    }
+
+    @Override
+    public void visitWhileStatement(WhileStatementTree tree) {
+      // do not scan the condition
+      addLines(tree.whileKeyword(), tree.closeParenToken());
+      addLineOfCloseBrace(tree.whileKeyword(), tree.statement());
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      // do not scan the condition
+      addLine(tree.doKeyword());
+      if (tree.doKeyword().line() != tree.whileKeyword().line()) {
+        addLines(tree.whileKeyword(), tree.semicolonToken());
+      }
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitForStatement(ForStatementTree tree) {
+      // do not scan the initializer, updater and condition
+      addLines(tree.forKeyword(), tree.closeParenToken());
+      addLineOfCloseBrace(tree.forKeyword(), tree.statement());
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitForEachStatement(ForEachStatement tree) {
+      // do not scan the variable and expression
+      addLines(tree.forKeyword(), tree.closeParenToken());
+      addLineOfCloseBrace(tree.forKeyword(), tree.statement());
+      scan(tree.statement());
+    }
+
+    @Override
+    public void visitTryStatement(TryStatementTree tree) {
+      // do not scan resources
+      if (tree.resources().isEmpty()) {
+        addLine(tree.tryKeyword());
+      } else {
+        addLines(tree.tryKeyword(), tree.closeParenToken());
+      }
+      scan(tree.block());
+      scan(tree.catches());
+      if (tree.finallyKeyword() != null) {
+        addLine(tree.finallyKeyword());
+      }
+      scan(tree.finallyBlock());
+    }
+
+    private void addLine(SyntaxToken token) {
+      statementsPerLine.add(token.line());
+    }
+
+    private void addLines(SyntaxToken startToken, SyntaxToken endToken) {
+      addLine(startToken);
+      if (startToken.line() != endToken.line()) {
+        addLine(endToken);
+      }
+    }
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    super.scanFile(context);
+
+    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(context.getFile());
+    // TODO minimize access to files, another one in LinesVisitor
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)